src/linear_algebra/matrix/diag.lean

```diff
@@ -0,0 +1,210 @@
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h]}
+end
+
+/-- Every unit matrix is diagonal. -/
+@[simp] lemma is_diag_unit [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+by {intros i j h, have h' := subsingleton.elim i j, contradiction}
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, by simp
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+by {intros i j h, simp [h]}
+
+@[simp] lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0):
+  (A.map f).is_diag :=
+by {intros i j h, simp [ha h, hf]}
+
+@[simp] lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by {intros i j h, simp [ha h]}
+
+@[simp] lemma is_diag.add
+[add_zero_class α] {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A + B).is_diag :=
+by {intros i j h, simp [ha h, hb h]}
+
+@[simp] lemma is_diag.sub [add_group α]
+{A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A - B).is_diag :=
+by {intros i j h, simp [ha h, hb h]}
+
+@[simp] lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
+{k : R} {A : matrix n n α} (ha : A.is_diag):
+  (k • A).is_diag :=
+by {intros i j h, simp [ha h]}
+
+@[simp] lemma is_diag.transpose
+[has_zero α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᵀ.is_diag :=
+by {intros i j h, simp [ha h.symm]}
+
+@[simp] lemma is_diag.conj_transpose
+[semiring α] [star_ring α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᴴ.is_diag :=
+by {intros i j h, simp [ha h.symm]}
+
+@[simp] lemma is_diag.minor [has_zero α]
+{A : matrix n n α} (ha : A.is_diag) {f : m → n} (hf : injective f):
+  (A.minor f f).is_diag:=
+by {intros i j h, specialize @hf i j, simp [ha (mt hf h)]}
+
+/-- `(A ⊗ B).is_diag` if both `A` and `B` are diagonal. -/
+@[simp] lemma is_diag.kronecker [mul_zero_class α]
+{A : matrix m m α} {B : matrix n n α} (hA: A.is_diag) (hB: B.is_diag):
+(A ⊗ₖ B).is_diag :=
+begin
+  rintros ⟨a, b⟩ ⟨c, d⟩ h,
+  dsimp [kronecker_apply],
+  by_cases hac: a = c,
+  { have hbd: b ≠ d, {tidy}, simp [hB hbd] },
+  { simp [hA hac] },
+end
+
+@[simp] lemma is_diag.diagonal [has_zero α] [decidable_eq n] (d : n → α) :
+  (diagonal d).is_diag :=
+by {intros i j h, simp [h]}
+
+lemma is_diag.is_symm [has_zero α] {A : matrix n n α} (h : A.is_diag) :
+  A.is_symm :=
+begin
+  ext,
+  by_cases g : i = j, {rw g},
+  simp [h g, h (ne.symm g)],
+end
+
+/-- The block matrix `A.from_blocks B C D` is diagonal
+    if  `A` and `D` are diagonal and `B` and `C` are `0`. -/
+lemma is_diag.from_blocks [has_zero α]
+  {A : matrix m m α} {B : matrix m n α} {C : matrix n m α} {D : matrix n n α}
+  (ha : A.is_diag) (hb : B = 0) (hc : C = 0) (hd : D.is_diag) :
+  (A.from_blocks B C D).is_diag:=
```

Comment:
I don't quite understand why we don't want `(hb : B = 0) (hc : C = 0)` here. I think we are more often to have the form `A.from_blocks B C D` and need to prove `(hb : B = 0) (hc : C = 0)` rather than having `A.from_blocks 0 0 D`, which does not seem very natural to me.

Reply:
That's what the `convert` tactic is for; it will automatically introduce `hb` and `hc` subgoals from my version of the lemma

Reply:
Hi, Eric. Are you sure the second version is better? If so, I will change it to the second one you gave.

Reply:
@eric-wieser

Reply:
You should have two versions; the one without B and C that I have in the suggestion above, and one with that is stated as an `iff`